biochemist-python/chapters/File_Parsing.ipynb

```diff
@@ -0,0 +1,602 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for line in data:\n",
+    "    if 'PROTEIN ATOMS' in line:\n",
+    "        PROTEIN_ATOM_line = line\n",
+    "        words = PROTEIN_ATOM_line.split()\n",
+    "        print(PROTEIN_ATOM_line)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that the fifth element in this list is a colon (:). It is possible to modify the split command to split lines using the colon as the delimiter (':')."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for line in data:\n",
+    "    if 'PROTEIN ATOMS' in line:\n",
+    "        PROTEIN_ATOM_line = line\n",
+    "        words = PROTEIN_ATOM_line.split(':')\n",
+    "        print(words)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "From this `print` statement, we now see that we have a list called words, where we have split `PROTEIN_ATOM_line`.  The number of atoms in the protein is actually the second element of this list, so we can now save it as a new variable."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "atoms = words[1]\n",
+    "print(atoms)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The HIV protease structure in 4eyr.pdb is a dimer. Let's find the number of atoms in each monomer unit. If we now try to do a math operation on atoms, we get an error message.  Why do you think that is?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "atoms / 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Even though `atoms` looks like a number to us, it is really a string, so we can not add an integer to it.  We need to change the data type of energy to a float. This is called *casting*."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "atoms = float(atoms)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now it will work.  If we thought ahead we could have changed the data type when we assigned the variable originally."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "atoms = float(words[1])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"exercise-this-is-a-title exercise\">\n",
+    "<p class=\"exercise-title\">Exercise on file parsing</p>\n",
```

Comment:
This exercise strikes me as being a jump from the material, and does not seem like something you would ask the students to do during the workshop (it's a lot of typing). Is this seen as a "homework" assignment/do you anticipate using it during a live workshop?

Reply:
In 4eyr.pdb, there is only one HETNAM line, representing the single non-water heterogroup in the structure, so this works with 4eyr.pdb. In a file such as 2w9g.pdb, there are three lines that start with HETNAM, representing three non-water heterogroups in the structure, so the three lines would print, but only the last one would be stored as HETNAM_line. I can’t think of a line that is guaranteed to occur only once in a PDB file. This is actually a flaw in the PDB file format and one of the reasons that the PDB has switched to cif format for their files. I think the example here serves the purpose of the notebook. If someone wants to dig deeper, they will encounter the problem you mentioned. Should I include a note on how to change the loop if there is more than one HETNAM line?